### PR TITLE
[NOGIL] Improve Test Coverage for Producer + Some Edge Case Fixes

### PR DESCRIPTION
**Pull request by ojasvajain**
1. Add integration test coverage for AIO Producer. This ensures AIO Producer works fine after all the changes done in Producer.c

2. Add more integration test cases related to Transactional Producer (using both sync and async Producer)

3. Addresses a TODO about calling abort_transaction after close has been initiated. A fix has been added to mimic Java client's behaviour by adding a abort_transaction call in close() itself. Ideally, the abort call should be made in the librdkafka layer but for the time being, it is added in the Python C bindings. With this fix in place, an abort_transaction call made after a close has been initiated will get an error (same as Java) but close will ensure it calls abort_transaction (same as Java) so calling abort explicitly will not be required.

4. Addresses a bug that allowed concurrent calls from inside a AIO consumer callback to be made. A lock has been introduced to serialize the concurrent calls. 

----

Checklist
------------------
- [ ] Contains customer facing changes? Including API/behavior changes 
- [ ] Did you add sufficient unit test and/or integration test coverage for this PR?
  - If not, please explain why it is not required

References
----------
JIRA: 

Test & Review
------------

Open questions / Follow-ups
--------------------------

### CODE REVIEW

File: src/confluent_kafka/aio/_common.py [hunk 1]
```diff
@@ -25,9 +26,10 @@
 T = TypeVar('T')
 
 
-class ReentryIdentity:
-    """Internal use only. The identity AIOConsumer presents to the Consumer
-    reentrancy gate.
+class ReentryContext:
```
Comment (k-raina) [inline src/confluent_kafka/aio/_common.py:31]: Two concurrent subscribe(), which double-frees callback references will be fixed by this fix but remains is the mutating call racing the application poll that is still running, still needs to be fixed?

eg. 
```
async def on_assign(consumer, partitions):
    asyncio.create_task(consumer.subscribe(...),)   
await consumer.subscribe(["topic"], on_assign=on_assign)
await consumer.poll(10)   
```

**Application workflow** 
1. poll() runs on executor worker W1,  holding the client handle rk.
2. librdkafka fires on_assign from inside that poll. W1 is now parked waiting for the callback to finish but its C-level poll is suspended mid-execution, still holding rk.
3. The callback starts consumer.subscribe(). Because asyncio copies the callback's context into the child task, the child also presents same identity.
4.. W2 runs Consumer_subscribe which runs/mutates on rk handle.
6. W1's suspended poll resumes and touches the mutated rk and crashs.

Can we check if this workflow should be fixed?

File: src/confluent_kafka/src/Producer.c [hunk 2]
```diff
@@ -625,6 +628,22 @@ Producer_close(Handle *self, PyObject *args, PyObject *kwargs) {
...
         CallState_begin(self, &cs);
 
+        /* If a transaction is in progress, abort it now rather than leaving it
+         * dangling on the broker. Ideally, rd_kafka_destroy() should do it on
+         * its own.
+         * _ERR__STATE (no transaction in progress) and _ERR__NOT_CONFIGURED
+         * (non-transactional producer) are the expected outcome for most
+         * producers and are ignored; anything else is stashed to warn about
+         * below. */
+        txn_error = rd_kafka_abort_transaction(self->rk, -1);
```
Comment (k-raina) [inline src/confluent_kafka/src/Producer.c:638]: -1 means an infinite deadline. librdkafka's abort ends with an EndTxn round-trip to the coordinator, which waits on that infinite deadline which blocks forever. Should we add resonable timeout?

File: src/confluent_kafka/aio/producer/_AIOProducer.py [hunk 3]
```diff
@@ -92,9 +92,9 @@ async def close(self) -> None:
         2. **Cancel Timeout Task**: Immediately cancels the buffer timeout monitoring task
         3. **Flush All Messages**: Flushes any buffered messages and waits for delivery confirmation
         4. **Shutdown ThreadPool**: Waits for all pending ThreadPool operations to complete
-        5. **Cleanup**: Ensures the underlying librdkafka producer is properly closed. The shutdown
-            is designed to be safe and non-blocking for the asyncio event loop
-            while ensuring all pending operations complete before the producer is closed.
+        5. **Cleanup** [Not implemented yet]: Ensures the underlying librdkafka producer is properly closed.
```
Comment (k-raina) [inline src/confluent_kafka/aio/producer/_AIOProducer.py:95]: Should we call  Producer.close() in AIOProducer.close()?

**Review by k-raina — CHANGES_REQUESTED** [review]
Couple of comments on correctness

File: src/confluent_kafka/aio/producer/_AIOProducer.py (hunk 3 above)
Comment (ojasvajain) [reply src/confluent_kafka/aio/producer/_AIOProducer.py:95]: Yes. It is a pre-existing bug that I am not planning to fix in this PR as it will require more testing.

File: src/confluent_kafka/src/Producer.c (hunk 2 above)
Comment (ojasvajain) [reply src/confluent_kafka/src/Producer.c:638]: In Java, when we call producer.close() without timeout, it blocks indefinitely on aborting the tx (if broker is not reachable), which is the what above change also does. But Java does provide an option to set timeout for close (which librdkafka does not support). Do you think setting 60s of timeout would be fine?

File: src/confluent_kafka/aio/_common.py (hunk 1 above)
Comment (ojasvajain) [reply src/confluent_kafka/aio/_common.py:31]: Yes this will break because task is outliving the callback. This can also happen in sync consumers if user starts some threads from within the callback. We should document that starting threads/tasks from within a callback is not recommended. I am not saying this can't be fixed but it can be taken as a future goal to fix maybe, if a requirement comes. In the current version of python client as well, the issue persists because there is no serialization at all so we are not breaking something that was supported. It's just that we are not supporting the serialization of consumer calls in all the cases, which should be fine.

File: src/confluent_kafka/src/Producer.c (hunk 2 above)
Comment (ojasvajain) [reply src/confluent_kafka/src/Producer.c:638]: To check with @emasab

File: src/confluent_kafka/aio/_common.py (hunk 1 above)
Comment (k-raina) [reply src/confluent_kafka/aio/_common.py:31]: Thanks. 
Lets verify this is issue with GIL mode also. If no, we might need to fix this in upcoming PRs.

File: src/confluent_kafka/aio/producer/_AIOProducer.py (hunk 3 above)
Comment (k-raina) [reply src/confluent_kafka/aio/producer/_AIOProducer.py:95]: Since rk is leaking after application is closed, seems like it will be one time leak so severity is low. We should check in future PRs if this can be fixed and free threading is compatible for fix.

**Review by k-raina — APPROVED** [review]
Thanks for PR @ojasvajain

File: src/confluent_kafka/src/Producer.c (hunk 2 above)
Comment (ojasvajain) [reply src/confluent_kafka/src/Producer.c:638]: On Emaneule's suggestion, I double checked the Java behaviour - the abort_tx call inside close waits indefinitely if broker is not reachable. So it should be fine to keep -1 as timeout here.

File: src/confluent_kafka/aio/_common.py (hunk 1 above)
Comment (ojasvajain) [reply src/confluent_kafka/aio/_common.py:31]: I pushed a fix for this. I have now introduced a new field called Invocation that will be set in context and instead of its value, its reference will be passed on to the child tasks. So when a callback is complete, it can update this object and then while executing the child tasks (who will also have the value updated), we can check if we want to treat it as a re-entrant call or a separate top-level call.

Comment (ojasvajain) [reply src/confluent_kafka/aio/_common.py:31]: Have reverted the fix based on Emanuele's suggestion. It will be part of a separate PR later.
